### PR TITLE
ddl,table/tables: handle partition by range column when len(column)=1

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

```
create table log_message_1(
add_time datetime not null default '2000-01-01 00:00:00',
log_level int unsigned not null default '0',
log_host varchar(32) not null,
service_name varchar(32) not null,
message varchar(2000)
)partition by range columns(add_time)(
partition p201403 values less than ('2014-04-01'),
partition p201404 values less than ('2014-05-01'),
partition p201405 values less than ('2014-06-01'),
partition p201406 values less than ('2014-07-01'),
partition p201407 values less than ('2014-08-01'),
partition p201408 values less than ('2014-09-01'),
partition p201409 values less than ('2014-10-01'),
partition p201410 values less than ('2014-11-01')
)
```

Fix error:

```
2019/01/15 17:06:59.808 partition.go:128: [error] wrong table partition expression:line 1 column 10 near " < ("2014-04-01"))" (total length 28)
github.com/pingcap/errors.AddStack
        /home/jenkins/workspace/build_tidb_2.1/go/pkg/mod/github.com/pingcap/errors@v0.11.0/errors.go:174
github.com/pingcap/errors.Trace
        /home/jenkins/workspace/build_tidb_2.1/go/pkg/mod/github.com/pingcap/errors@v0.11.0/juju_adaptor.go:12
github.com/pingcap/parser.(*Parser).Parse
        /home/jenkins/workspace/build_tidb_2.1/go/pkg/mod/github.com/pingcap/parser@v0.0.0-20181207085916-6c21d4344dfa/yy_parser.go:111
github.com/pingcap/tidb/expression.ParseSimpleExprWithTableInfo
        /home/jenkins/workspace/build_tidb_2.1/go/src/github.com/pingcap/tidb/expression/simple_rewriter.go:40
github.com/pingcap/tidb/table/tables.generatePartitionExpr
        /home/jenkins/workspace/build_tidb_2.1/go/src/github.com/pingcap/tidb/table/tables/partition.go:125
github.com/pingcap/tidb/table/tables.newPartitionedTable
        /home/jenkins/workspace/build_tidb_2.1/go/src/github.com/pingcap/tidb/table/tables/partition.go:66
github.com/pingcap/tidb/table/tables.TableFromMeta
        /home/jenkins/workspace/build_tidb_2.1/go/src/github.com/pingcap/tidb/table/tables/tables.go:139
github.com/pingcap/tidb/infoschema.(*Builder).applyCreateTable
        /home/jenkins/workspace/build_tidb_2.1/go/src/github.com/pingcap/tidb/infoschema/builder.go:178
github.com/pingcap/tidb/infoschema.(*Builder).ApplyDiff
        /home/jenkins/workspace/build_tidb_2.1/go/src/github.com/pingcap/tidb/infoschema/builder.go:94
github.com/pingcap/tidb/domain.(*Domain).tryLoadSchemaDiffs
        /home/jenkins/workspace/build_tidb_2.1/go/src/github.com/pingcap/tidb/domain/domain.go:231
github.com/pingcap/tidb/domain.(*Domain).loadInfoSchema
        /home/jenkins/workspace/build_tidb_2.1/go/src/github.com/pingcap/tidb/domain/domain.go:104
github.com/pingcap/tidb/domain.(*Domain).Reload
        /home/jenkins/workspace/build_tidb_2.1/go/src/github.com/pingcap/tidb/domain/domain.go:313
github.com/pingcap/tidb/domain.(*Domain).loadSchemaInLoop
```

### What is changed and how it works?

Add preliminary support for `partition by range column`
'create table ... partition by range column (c)'
can be transformed to 'create table ... partiton by range c'

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

Side effects

With the introduction of this new partition type, we have to check more things:

1. Check value constraint for `create table ... partition by column`
2. Check value constraint for `alter table add partition ...`
3. Add CRUD test for partition by column(c)

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @zimulala

**Comment by tiancaiamao** [comment]
PTAL @zimulala @winkyao @ciscoxll

File: ddl/partition.go [hunk 1]
```diff
@@ -65,6 +65,10 @@ func buildTablePartitionInfo(ctx sessionctx.Context, d *ddl, s *ast.CreateTableS
 		s.Partition.Expr.Format(buf)
 		pi.Expr = buf.String()
 	} else if s.Partition.ColumnNames != nil {
+		// TODO: Support multiple columns for 'partition by range column'.
```
Comment (ciscoxll) [inline ddl/partition.go]: s / 'partition by range column' / 'partition by range columns'

File: ddl/ddl.go [hunk 2]
```diff
@@ -94,6 +94,9 @@ var (
 
 	errUnsupportedShardRowIDBits = terror.ClassDDL.New(codeUnsupportedShardRowIDBits, "unsupported shard_row_id_bits for table with auto_increment column.")
 
+	errUnsupportedPartitionByRangeColumn = terror.ClassDDL.New(codeUnsupportedPartitionByRangeColumn,
+		"unsupported partition by range column")
```
Comment (ciscoxll) [inline ddl/ddl.go]: unsupported multiple columns for 'partition by range columns'?

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -1192,6 +1191,19 @@ func checkPartitionByRange(ctx sessionctx.Context, tbInfo *model.TableInfo, pi *
 	return nil
 }
 
+func checkPartitionByRangeColumn(ctx sessionctx.Context, tbInfo *model.TableInfo, pi *model.PartitionInfo, s *ast.CreateTableStmt) error {
+	if err := checkPartitionNameUnique(tbInfo, pi); err != nil {
+		return err
+	}
+
+	// TODO: Check CreatePartitionValue.
```
Comment (ciscoxll) [inline ddl/ddl_api.go]: Range columns partition key supports multiple data types with integer、datetime、string.

Comment (tiancaiamao) [reply ddl/ddl_api.go]: Yes, I know, and I'll do it later.
https://github.com/pingcap/tidb/issues/9083

**Comment by tiancaiamao** [comment]
PTAL @ciscoxll @zimulala

**Comment by ciscoxll** [comment]
PTAL @zimulala @crazycs520

File: ddl/partition.go [hunk 4]
```diff
@@ -65,6 +65,10 @@ func buildTablePartitionInfo(ctx sessionctx.Context, d *ddl, s *ast.CreateTableS
 		s.Partition.Expr.Format(buf)
 		pi.Expr = buf.String()
 	} else if s.Partition.ColumnNames != nil {
+		// TODO: Support multiple columns for 'PARTITION BY RANGE COLUMNS'.
+		if enable && len(s.Partition.ColumnNames) != 1 {
+			return nil, errors.Trace(errUnsupportedPartitionByRangeColumn)
```
Comment (crazycs520) [inline ddl/partition.go]: Please add tests for this.

Comment (tiancaiamao) [reply ddl/partition.go]: Done @crazycs520

**Comment by tiancaiamao** [comment]
/run-all-tests
PTAL @crazycs520  @winkyao

File: ddl/ddl_api.go (hunk 3 above)
Comment (zimulala) [reply ddl/ddl_api.go]: Do we need to check `validRangePartitionType`?

File: ddl/db_partition_test.go [hunk 5]
```diff
@@ -257,6 +266,42 @@ func (s *testIntegrationSuite) TestCreateTableWithHashPartition(c *C) {
...
+) partition by range columns(add_time)(
+    partition p201403 values less than ('2014-04-01'),
+    partition p201404 values less than ('2014-05-01'),
+    partition p201405 values less than ('2014-06-01'),
+    partition p201406 values less than ('2014-07-01'),
+    partition p201407 values less than ('2014-08-01'),
+    partition p201408 values less than ('2014-09-01'),
+    partition p201409 values less than ('2014-10-01'),
+    partition p201410 values less than ('2014-11-01')
+)`)
```
Comment (zimulala) [inline ddl/db_partition_test.go]: Insert some records and do "select * from log_message_1"?

Comment (tiancaiamao) [reply ddl/db_partition_test.go]: 1. This function just `TestCreateTableWithRangeColumnPartition`, not CRUD
2. Actually I manually test some DML statements
3. I should have port some mysql test for partition if this is not a special week and table partition have the lower priority. Porting mysql test will cover here.

File: ddl/ddl_api.go (hunk 3 above)
Comment (tiancaiamao) [reply ddl/ddl_api.go]: Range partition type is checked.
This is the new added range columns partition.

File: ddl/table.go [hunk 6]
```diff
@@ -682,7 +682,7 @@ func updatePartitionInfo(partitionInfo *model.PartitionInfo, tblInfo *model.Tabl
 
 // checkAddPartitionValue values less than value must be strictly increasing for each partition.
 func checkAddPartitionValue(meta *model.TableInfo, part *model.PartitionInfo) error {
-	if meta.Partition.Type == model.PartitionTypeRange {
+	if meta.Partition.Type == model.PartitionTypeRange && len(meta.Partition.Columns) == 0 {
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply ddl/table.go]: There is already a `TODO`, if we uncomment
https://github.com/pingcap/tidb/pull/9082/files#diff-edfa54cbe2fdb418f0a77bd0ec4a56f2R1286

We certainly find it's necessary to update here, so an additional TODO is not necessary.

File: ddl/ddl_api.go [hunk 7]
```diff
@@ -2688,13 +2701,18 @@ func buildPartitionInfo(meta *model.TableInfo, d *ddl, spec *ast.AlterTableSpec)
...
+					if strings.EqualFold(buf.String(), "MAXVALUE") {
+						continue
+					}
+					buf.Reset()
+					return nil, infoschema.ErrColumnNotExists.GenWithStackByArgs(buf.String(), "partition function")
 				}
-				buf.Reset()
-				return nil, infoschema.ErrColumnNotExists.GenWithStackByArgs(buf.String(), "partition function")
+			} else {
+				// Partition by range column.
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply ddl/ddl_api.go]: Because we need the commit

[the remark this answers is not in the document]
Comment (winkyao) [reply ddl/ddl_api.go]: You can comment the `else` too.

[the remark this answers is not in the document]
Comment (tiancaiamao) [reply ddl/ddl_api.go]: Done

**Comment by tiancaiamao** [comment]
PTAL @crazycs520

**Comment by tiancaiamao** [comment]
PTAL @zimulala

**Comment by tiancaiamao** [comment]
PTAL @zimulala

**Comment by tiancaiamao** [comment]
PTAL @zimulala

**Pull request merged**